File: auto_outsrc/parser/LW/LW_LATEST.cpp
#include "Charm.h"
#include <iostream>
#include <sstream>
#include <string>
#include <list>
using namespace std;

PairingGroup group(AES_SECURITY);

SecretUtil util;

void setup(CharmList & gpk)
{
    G1 g = group.init(G1_t);
    g = group.random(G1_t);
    gpk.insert(0, g);
    return;
}

void authsetup(CharmList & gpk, CharmListStr & authS, CharmMetaList & msk, CharmMetaList & pk)
{
    G1 g;
    int Y = 0;
    ZR alpha = group.init(ZR_t);
    ZR y = group.init(ZR_t);
    string z;
    GT eggalph = group.init(GT_t);
    G1 gy = group.init(G1_t);
    CharmList tmpList0;
    CharmList tmpList1;
    
    g = gpk[0].getG1();
    Y = authS.length();
    for (int i = 0; i < Y; i++)
    {
        alpha = group.random(ZR_t);
        y = group.random(ZR_t);
        z = authS[i];
        eggalph = group.exp(group.pair(g, g), alpha);
        gy = group.exp(g, y);
        tmpList0.insert(0, alpha);
        tmpList0.insert(1, y);
        msk.insert(z, tmpList0);
        tmpList1.insert(0, eggalph);
        tmpList1.insert(1, gy);
        pk.insert(z, tmpList1);
    }
    return;
}

void keygen(CharmList & gpk, CharmMetaList & msk, string & gid, CharmListStr & userS, CharmListZR & blindingFactorKBlinded, CharmList & skBlinded)
{
    G1 g;
    G1 h = group.init(G1_t);
    int Y = 0;
    string z;
    CharmListG1 K;
    CharmListG1 KBlinded;
    
    g = gpk[0].getG1();
    h = group.hashListToG1(gid);
    Y = userS.length();
    for (int i = 0; i < Y; i++)
    {
        z = userS[i];
        K.insert(z, group.mul(group.exp(g, msk[z][0].getZR()), group.exp(h, msk[z][1].getZR())));
    }
    CharmList K_Keys_List = K.keys();
    int K_List_Length = K.length();
    for (int y_Temp_Loop_Var = 0; y_Temp_Loop_Var < K_List_Length; y_Temp_Loop_Var++)
    {
        y = K_Keys_List[y_Temp_Loop_Var];
        blindingFactorKBlinded.insert(y, group.random(ZR_t));
        KBlinded.insert(y, group.exp(K[y], group.div(1, blindingFactorKBlinded[y])));
    }
    skBlinded.insert(0, gid);
    skBlinded.insert(1, KBlinded);
    return;
}

void encrypt(CharmMetaList & pk, CharmList & gpk, GT & M, string & policy_str, CharmList & ct)
{
    G1 g;
    Policy policy;
    CharmListStr attrs;
    ZR s = group.init(ZR_t);
    int w = 0;
    CharmDictZR s_sh;
    CharmDictZR w_sh;
    int Y = 0;
    GT egg = group.init(GT_t);
    GT C0 = group.init(GT_t);
    ZR r = group.init(ZR_t);
    string k;
    CharmListGT C1;
    CharmListG1 C2;
    CharmListG1 C3;
    
    g = gpk[0].getG1();
    policy = util.createPolicy(policy_str);
    attrs = util.getAttributeList(policy);
    s = group.random(ZR_t);
    w = 0;
    s_sh = util.calculateSharesDict(group, s, policy);
    w_sh = util.calculateSharesDict(group, w, policy);
    Y = s_sh.length();
    egg = group.pair(g, g);
    C0 = group.mul(M, group.exp(egg, s));
    for (int y = 0; y < Y; y++)
    {
        r = group.random(ZR_t);
        k = attrs[y];
        C1.insert(k, group.mul(group.exp(egg, s_sh[k]), group.exp(pk[k][0].getGT(), r)));
        C2.insert(k, group.exp(g, r));
        C3.insert(k, group.mul(group.exp(pk[k][1].getG1(), r), group.exp(g, w_sh[k])));
    }
    ct.insert(0, policy_str);
    ct.insert(1, C0);
    ct.insert(2, C1);
    ct.insert(3, C2);
    ct.insert(4, C3);
    return;
}

void transform(CharmList & skBlinded, CharmListStr & userS, CharmList & ct, CharmList & transformOutputList, CharmDictZR & coeff, int Y)
{

File: auto_outsrc/parser/LW/LW_LATEST.py
from LW_LATEST_USER import *

from charm.toolbox.pairinggroup import PairingGroup,ZR,G1,G2,GT,pair
from charm.core.engine.util import *
from charm.core.math.integer import randomBits

group = None

secparam = 80


def setup():
    g = group.random(G1)
    gpk = [g]
    output = gpk
    return output

def authsetup(gpk, authS):
    msk = {}
    pk = {}

    [g] = gpk
    Y = len(authS)
    for i in range(0, Y):
        alpha = group.random(ZR)
        y = group.random(ZR)
        z = authS[i]
        eggalph = (pair(g, g) ** alpha)
        gy = (g ** y)
        msk[z] = [alpha, y]
        pk[z] = [eggalph, gy]
    output = (msk, pk)
    return output

def keygen(gpk, msk, gid, userS):
    K = {}
    KBlinded = {}
    blindingFactorKBlinded = {}

    [g] = gpk
    h = group.hash(gid, G1)
    Y = len(userS)
    for i in range(0, Y):
        z = userS[i]
        K[z] = ((g ** msk[z][0]) * (h ** msk[z][1]))
    for y in K:
        blindingFactorKBlinded[y] = group.random(ZR)
        KBlinded[y] = (K[y] ** (1 / blindingFactorKBlinded[y]))
    skBlinded = [gid, KBlinded]
    output = (blindingFactorKBlinded, skBlinded)
    return output

def encrypt(pk, gpk, M, policy_str):
    C2 = {}
    C1 = {}
    C3 = {}

    [g] = gpk
    policy = createPolicy(policy_str)
    attrs = getAttributeList(policy)
    s = group.random(ZR)
    w = 0
    s_sh = calculateSharesDict(s, policy)
    w_sh = calculateSharesDict(w, policy)
    Y = len(s_sh)
    egg = pair(g, g)
    C0 = (M * (egg ** s))
    for y in range(0, Y):
        r = group.random(ZR)
        k = attrs[y]
        C1[k] = ((egg ** s_sh[k]) * (pk[k][0] ** r))
        C2[k] = (g ** r)
        C3[k] = ((pk[k][1] ** r) * (g ** w_sh[k]))
    ct = [policy_str, C0, C1, C2, C3]
    output = ct
    return output

def transform(skBlinded, userS, ct):
    transformOutputList = {}

    policy_str, C0, C1, C2, C3 = ct
    gidBlinded, KBlinded = skBlinded
    transformOutputList[1] = C0
    policy = createPolicy(policy_str)
    attrs = prune(policy, userS)
    coeff = getCoefficients(policy)
    transformOutputList[0] = group.hash(gidBlinded, G1)
    h_gid = transformOutputList[0]
    Y = len(attrs)
    for y in range(0, Y):
        pass
        kDecrypt = GetString(attrs[y])
        transformOutputList[1000+12*y] = pair(h_gid, C3[kDecrypt])
        result0 = transformOutputList[1000+12*y]
        transformOutputList[1001+12*y] = ((result0 ** coeff[kDecrypt]) * (C1[kDecrypt] ** coeff[kDecrypt]))
        numerator = transformOutputList[1001+12*y]
        transformOutputList[1002+12*y] = pair(KBlinded[kDecrypt], C2[kDecrypt])
        denominator0 = transformOutputList[1002+12*y]
    output = (transformOutputList, coeff, Y, attrs)
    return output

def decout(userS, transformOutputList, blindingFactorKBlinded, coeff, Y, attrs):
    C0 = transformOutputList[1]
    h_gid = transformOutputList[0]
    dotProd = group.init(GT)
    for y in range(0, Y):
        pass
        kDecrypt = GetString(attrs[y])
        result0 = transformOutputList[1000+12*y]
        numerator = transformOutputList[1001+12*y]
        denominator0 = (transformOutputList[1002+12*y] ** blindingFactorKBlinded[kDecrypt])
        denominator = (denominator0 ** coeff[kDecrypt])
        fraction = (numerator * (denominator ** -1))
        dotProd = (dotProd * fraction)
    M = (C0 * (dotProd ** -1))
    output = M
    return output

def SmallExp(bits=80):
    return group.init(ZR, randomBits(bits))

def main():
    global group
    #group = PairingGroup(secparam)

    group = PairingGroup("SS512")

    gpk = setup()
    (msk, pk) = authsetup(gpk, ['ONE', 'TWO', 'THREE', 'FOUR'])
    (blindingFactorKBlinded, skBlinded) = keygen(gpk, msk, "john@example.com", ['ONE', 'TWO', 'THREE'])
    M = group.random(GT)
    ct = encrypt(pk, gpk, M, '((four or three) and (two or one))')
    (transformOutputList, coeff, Y, attrs) = transform(skBlinded, ['ONE', 'TWO', 'THREE'], ct)
    M2 = decout(['ONE', 'TWO', 'THREE'], transformOutputList, blindingFactorKBlinded, coeff, Y, attrs)
    print(M)
    print("\n\n\n")
    print(M2)
    print("\n\n")
    if (M == M2):
        print("success")
    else:
        print("failed")


if __name__ == '__main__':
    main()

